kernel/kernel-api/src/main/java/io/delta/kernel/coordinatedcommits/AbstractCommitCoordinatorBuilder.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.coordinatedcommits;
+
+import io.delta.kernel.annotation.Evolving;
+import io.delta.kernel.config.ConfigurationProvider;
+import io.delta.kernel.internal.DeltaErrors;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * A builder class to create a {@link CommitCoordinatorClient} (CCC).
+ *
+ * <p>Subclasses of this class must provide a no-argument constructor.
+ *
+ * <p>Note: These builder classes are optional. The {@link
+ * io.delta.kernel.engine.Engine#getCommitCoordinatorClient} API does not prescribe how to create
+ * the underlying CCC and does not require a builder.
+ *
+ * <p>The benefit of implementing a builder for your CCC is that your Engine may then invoke {@link
```

Comment:
What's the alternative solution for a CC if a CC builder is not implemented? Can you please specify that too? In general, as an engine owner, how do I decide whether I should implement a builder or not?

Reply:
The alternative is: build it yourself! Who knows what clients they are instantiating. Maybe those clients have buildings that they want to be instantiated via reflection. Maybe some engine hardcodes what client to use and there's no need to have a fancy builder at all.

It's completely up to engines.

> how do I decide whether I should implement a builder or not?
When you want the CCC instantaited at runtime via reflection. i.e. you want which CCC to use to be decided at runtime. I can add this.